doc/developer/design/20211126-microbenchmarks.md

```diff
@@ -0,0 +1,135 @@
+# Microbenchmarking framework
+
+## Summary
+
+In order to characterize the performance of the product and guard against regressions, we need to be able to benchmarks the invidiual operations Mz performs when it comes to ingesting and processing data and materializing views.
+
+Microbenchmarking will be in addition to any end-to-end benchmarks that exist or will be developed. It aims to provide a more focused spotlight on regressions: instead of a 2% regression in an entire industry benchmark between product versions X and Y, we will be looking at a 20% regression in a specific named operation between specific commits M and N.
```

Comment:
So I think of something a bit different when I hear the term “microbenchmark.” Specifically I think of the criterion.rs benchmarks that we have scattered about that test a single Rust function in isolation.

I don’t have a great word for the benchmarks proposed here, but maybe something like “feature benchmark”? Like, they still test the entire product end-to-end, no? It’s just that we’re hoping to time the execution of just one feature from the outside.

Reply:
Yes, "feature benchmark" seems about right! Will use it going forward.